keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,537 @@
+}
+```
+
+Here is the definition of the ExecutionHookSpec:
+
+```
+// HookActionName is copied to ExecutionHookSpec by the controller such as
+// the Snapshot Controller.
+type ExecutionHookSpec struct {
+        // PodSelection defines how to select pods and containers to run
+	// the executionhook. If multiple pod/containers are selected, the action will executed on them
+	// asynchronously. If execution ordering is required, caller has to implement the logic and create
+	// different hooks in order.
+	// This field is required.
+        PodSelection PodSelection
+
+        // Name of the HookAction. This is required.
+        ActionName string
+}
+
+// PodSelection contains two fields, PodContainerNamesList and PodContainerSelector,
+// where one of them must be defined so that the hook controller knows where to
+// run the hook.
+type PodSelection struct {
+        // PodContainerNamesList lists the pods/containers on which the ExecutionHook
+        // should be executed. If not specified, the ExecutionHook controller will find
+        // all pods and containers based on PodContainerSelector.
+        // If both PodContainerNamesList and PodContainerSelector are not
+        // specified, the ExecutionHook cannot be executed and it will fail.
+        // +optional
+        PodContainerNamesList []PodContainerNames
+
+        // PodContainerSelector is for hook controller to find pods and containers
+        // based on the pod label selector and container names
+        // If PodContainerNamesList is specified, this field will not be used.
+        // +optional
+        PodContainerSelector *PodContainerSelector
+}
+
+type PodContainerNames struct {
+        PodName string
+        ContainerNames []string
+}
+
+type PodContainerSelector struct {
+	// PodSelector specifies a label query over a set of pods.
+        // +optional
+	PodSelector *metav1.LabelSelector
+
+        // If specified, controller only select the containers that are listed from the selected pods based on PodSelector. 
+	// Otherwise, all containers of the pods will be selected
+        // +optional
+        ContainerList []string
+ }
+```
+
+Here is the definition of the ExecutionHookStatus. This represents the current state of a hook for all selected containers in the selected pods.
+
+```
+// ExecutionHookStatus represents the current state of a hook
+type ExecutionHookStatus struct {
+        // This is a list of ContainerExecutionHookStatus, with each status representing information
+        // about how hook is executed in a container, including pod name, container name, ActionTimestamp,
+        // ActionSucceed, etc.
+        // +optional
+        HookStatuses []ContainerExecutionHookStatus
+
+        // Action Summary status
+        // Default is nil
+        // +optional
+        ActionSucceed *bool
+}
+
+// ContainerExecutionHookStatus represents the current state of a hook for a specific container in a pod
+type ContainerExecutionHookStatus struct {
```

Comment:
to revisit a comment, this will replicate PodName if there are multiple containers in the pod.  Is it worth structuring as 

```
{
    podName: pod-1234
    containers:
        - name: app
          timestamp: 12345678
          success: true
        - name: sidecar
          success: false
```

?